test/unit/hyperloglog.c

```diff
@@ -0,0 +1,392 @@
+	}
+
+	double avg_std_err = std_err_sum / n_points;
+	double avg_max_err = max_err_sum / n_points;
+
+	res->std_err = avg_std_err;
+	res->max_err = avg_max_err;
+}
+
+void
+measure_sparse_hll_estimation_error(int prec, size_t max_card,
+				    size_t n_points, size_t sets_per_point,
+				    struct est_errors *res, FILE *output)
+{
+	measure_hll_estimation_error(prec, HLL_SPARSE, max_card,
+				     n_points, sets_per_point, res, output);
+}
+
+void
+measure_dense_hll_estimation_error(int prec, size_t max_card,
+				   size_t n_points, size_t sets_per_point,
+				   struct est_errors *res, FILE *output)
+{
+	measure_hll_estimation_error(prec, HLL_DENSE, max_card,
+				     n_points, sets_per_point, res, output);
+}
+
+/*
+ * This test can dump the data that is used to measure
+ * the estimation error. These data can be used for further
+ * analysis and empirical based impovements of the algorithm.
+ */
+void
+test_dense_hyperloglog_error(void)
+{
+	header();
+	plan(HLL_N_PRECISIONS);
+
+	struct est_errors errors[HLL_MAX_PRECISION + 1];
+
+	FILE *output = NULL;
+	if (OUTPUT_FILE_NAME != NULL) {
+		output = fopen(OUTPUT_FILE_NAME, "w");
+		assert(output && "Can't open the output file.");
+	}
+
+	MAYBE_PRINT(output, "prec, card, avg_est, std_err, max_err\n");
+
+	for (int prec = HLL_MIN_PRECISION;
+	     prec <= HLL_MAX_PRECISION; ++prec) {
+		/*
+		 * ** let m = n_registers **
+		 * The range of 10m is divided into 3 sections with different
+		 * estimation approach, each of which must be tested:
+		 * [0 - m] - LinearCounting algorithm.
+		 * [1m - 5m] - HyperLogLog with bias correction.
+		 * [5m - 10m] - pure HyperLogLog algorithm.
+		 */
+		size_t max_card = 10 * n_registers(prec);
+		measure_dense_hll_estimation_error(prec, max_card,
+						   N_POINTS, SETS_PER_POINT,
+						   errors + prec, output);
+		/*
+		 * The error of HyperLogLog is close to 1/sqrt(n_counters),
+		 * but for small cardinalities LinearCounting is used because
+		 * it has better accuracy, so the resulting error must be
+		 * smaller than the HyperLogLog theoretical error.
+		 */
+		ok(errors[prec].std_err < hll_error(prec),
+		   "The actual error doesn't exceed the expected value.");
+	}
+
+	if (output)
+		fflush(output);
+
+	for (int prec = HLL_MIN_PRECISION;
+	     prec <= HLL_MAX_PRECISION; ++prec) {
+		MAYBE_PRINT(output,
+			    "prec:%d, std_err:%lg, max_err:%lg, exp_err: %lg\n",
+			    prec, errors[prec].std_err, errors[prec].max_err,
+			    hll_error(prec));
+	}
+
+	check_plan();
+	footer();
+}
+
+void
+test_sparse_hyperloglog_error(void)
+{
+	header();
+	/*
+	 * Since the precision parameter defines only the maximal size of the
+	 * sparse representation there is no need in tests for all available
+	 * precision values.
+	 */
+	plan(1);
+
+	int dprec = HLL_MAX_PRECISION;
+	struct est_errors errors;
+	/*
+	 * Sparse representation stores 4-byte pairs instead of 6-bit counters,
+	 * so 6/32 < 6 times fewer pairs can be stored in the same amount of
```

Comment:
```suggestion
	 * so 32/6 < 6 times fewer pairs can be stored in the same amount of
```

If this really is a typo, please fix it everywhere.

Reply:
Fixed the typos.